cpu/nrf52/radio/nrf802154/nrf802154.c

```diff
@@ -0,0 +1,403 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_nrf52
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the radio driver for nRF52 radios
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "luid.h"
+#include "mutex.h"
+
+#include "net/ieee802154.h"
+#include "net/netdev/ieee802154.h"
+#include "nrf802154.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static const netdev_driver_t nrf802154_netdev_driver;
+
+netdev_ieee802154_t nrf802154_dev = {
+    {
+        .driver = &nrf802154_netdev_driver,
+        .event_callback = NULL,
+        .context = NULL,
+    #ifdef MODULE_NETSTATS_L2
+        .stats = { 0 }
+    #endif
+    },
+#ifdef MODULE_GNRC
+#ifdef MODULE_GNRC_SIXLOWPAN
+    .proto = GNRC_NETTYPE_SIXLOWPAN,
+#else
+    .proto = GNRC_NETTYPE_UNDEF,
+#endif
+#endif
+    .pan = IEEE802154_DEFAULT_PANID,
+    .short_addr = { 0, 0 },
+    .long_addr = { 0, 0, 0, 0, 0, 0, 0, 0 },
+    .chan = IEEE802154_DEFAULT_CHANNEL,
+    .flags = 0
+};
+
+static uint8_t rxbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* length PHR + PSDU + LQI */
+static uint8_t txbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* length PHR + PSDU + LQI */
+
+static mutex_t txlock;
+
+/**
+ * @brief   Set radio into DISABLED state
+ */
+static void disable(void)
+{
+    /* set device into DISABLED state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {
+        NRF_RADIO->EVENTS_DISABLED = 0;
+        NRF_RADIO->TASKS_DISABLE = 1;
+        while (!(NRF_RADIO->EVENTS_DISABLED)) {}
+        DEBUG("Device state: DISABLED\n");
+    }
+}
+
+/**
+ * @brief   Set radio into RXIDLE state
+ */
+static void enable_rx(void)
+{
+    DEBUG("Set device state to RXIDLE\n");
+    NRF_RADIO->PACKETPTR = (uint32_t)rxbuf;
+    /* set device into RXIDLE state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_TxIdle) {
+        disable();
+    }
+    NRF_RADIO->EVENTS_RXREADY = 0;
+    NRF_RADIO->TASKS_RXEN = 1;
+    while (!(NRF_RADIO->EVENTS_RXREADY)) {}
+    DEBUG("Device state: RXIDLE\n");
+}
+
+/**
+ * @brief   Set radio into TXIDLE state
+ */
+static void enable_tx(void)
+{
+    DEBUG("Set device state to TXIDLE\n");
+    NRF_RADIO->PACKETPTR = (uint32_t)txbuf;
+    /* set device into TXIDLE state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_RxIdle) {
+        disable();
+    }
+    NRF_RADIO->EVENTS_TXREADY = 0;
+    NRF_RADIO->TASKS_TXEN = 1;
+    while (!(NRF_RADIO->EVENTS_TXREADY)) {}
+    DEBUG("Device state: TXIDLE\n");
+}
+
+/**
+ * @brief   Reset the RXIDLE state
+ */
+ static void reset_rx(void)
+ {
+     /* reset RX buffer and listen for new packets */
+     rxbuf[0] = 0;
+     NRF_RADIO->TASKS_START = 1;
+ }
+
+/**
+ * @brief   Switch on the radio
+ */
+static void power_on(void)
```

Comment:
you could make this function explicitly inline (but the compile does this probably anyway, so its rather one for the eye...)